### PR TITLE
CIP-0137? | Decentralized Message Queue

### PR DESCRIPTION
**Pull request by jpraynaud**
This proposal specifies a decentralized message queue leveraging the Cardano network layer. This protocol allows a topic based mechanism to diffuse messages from publishers to consumers in a decentralized way.

([rendered proposal](https://github.com/cardano-scaling/CIPs/blob/decentralized-message-queue/CIP-0137/README.md))

### CODE REVIEW

File: CIP-0XXX/README.md
Comment (Ryun1) [inline CIP-0XXX/README.md:22]:
```suggestion
```
the new template is held at https://github.com/cardano-foundation/CIPs/blob/master/.github/CIP-TEMPLATE.md

please follow

Comment (rphair) [reply CIP-0XXX/README.md:22]: @jpraynaud there are several suggestions, with many examples, for updating the old to the new format here (you are using kind of a mix of the old & new formats):
- #389 

If you have difficulty getting started with that, just post here saying so and I'll contribute a review with the necessary formatting changes.

File: CIP-0XXX/README.md
Comment (ch1bo) [reply CIP-0XXX/README.md:4]: > if all the network behaviour described herein is in the consensus layer

Categorizing this CIP will be interesting and we will likely encounter this scoping discussion several times throughout review of this CIP:

We saw three protocols with a need for this CIP: Mithril, Praos and Leios. Only the latter two are scoped to be used by the cardano node itself for consensus. In any case, the goal of this CIP is not to define the consensus protocols, but rather to standardize the network protocol used for message submission.

Comment (abailly) [reply CIP-0XXX/README.md:4]: If anything, it seems to me the right category for this CIP should be `Network`

Comment (rphair) [reply CIP-0XXX/README.md:4]: @abailly the CIP editors would be really happy to see a `Network` category: especially if you & your team would also identify any previous CIPs & drafts which would be eligible for it.

`Consensus` is proceeding as a CIP category due to the work that @dnadales mentioned here (as far as I know, the terms "enroll" and "enlist" are being used interchangeably):
- https://github.com/cardano-foundation/CIPs/pull/872#discussion_r1706965548
- https://github.com/IntersectMBO/ouroboros-consensus/issues/1205

... of documenting it with a CIP and fulfilling other expectations of [enlistment](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0001/README.md#project-enlisting).  We would love to hear that something similar could happen / is happening for a new `Network` category.

Comment (coot) [reply CIP-0XXX/README.md:4]: We shouldn't be held by review of all the previous CIPs, but the core network protocols haven't had any CIP so far, at least (co-)authored by the networking team.  It could be that there was a CIP to add the `LocalTxMonitor` mini-protocol (done in this PR IntersectMBO/ouroboros-network#3404 by @KtorZ) - that would be a good candidate if there was a CIP for that.

Comment (abailly) [reply CIP-0XXX/README.md:4]: @coot Of course, what happened happened. I am only suggesting that at least this CIP, and possibly others to come in the future, would be better categorised as `Network` and therefore it would be great if there was such a category, with some rules and guidelines.

Comment (rphair) [reply CIP-0XXX/README.md:4]: > We shouldn't be held by review of all the previous CIPs

@coot I wasn't suggesting that you be obligated to review them: only that CIP Editors would want to identify any "Network" CIPs you could point out, and likely assign a `Network` category to those retroactively if & when it becomes an enlisted category.

Comment (coot) [reply CIP-0XXX/README.md:4]: If you identify such list, I can review it if you'd need some help.

File: CIP-0XXX/README.md
Comment (ch1bo) [reply CIP-0XXX/README.md:46]: Notably, this diagram does _not_ say that "network node" would be the cardano node. It could be any node implementing this protocol, even a "mithril network node" as you have sketched it out in that "# Information Diffusion Architecture" section.

File: CIP-0XXX/README.md
Comment (ch1bo) [reply CIP-0XXX/README.md:182]: Yes. This step only needs to be done when the KES key changes.

**Comment by coot** [comment]
I'd suggest a different title for the PR/CIP; what about `Mithril over Ouroboros-Network`?

**Comment by rphair** [comment]
@coot (also responding to https://github.com/cardano-foundation/CIPs/pull/876#discussion_r1713603981) it depends on whether there will someday be another application of this method besides the one being developed for Mithril.

Will that be possible?  From my level of understanding (not intimate) that possibility is suggested in the _Motivation_:
> probably other protocols in the future

File: CIP-0XXX/README.md
Comment (rphair) [inline CIP-0XXX/README.md:538]:
```suggestion
## Copyright
```
(continued)

File: CIP-0XXX/README.md
Comment (rphair) [inline CIP-0XXX/README.md:13]:
```suggestion
```
We would have to drop this in the editing process... during which it's generally understood that additional authors can be added.

File: CIP-0XXX/README.md
Comment (rphair) [inline CIP-0XXX/README.md:17]:
```suggestion
    - https://github.com/cardano-foundation/CIPs/pull/876
```
(this discussion)

**Review by rphair — COMMENTED** [review]
Some formatting changes that would be necessary to comply with CIP-0001 as mentioned here: https://github.com/cardano-foundation/CIPs/pull/876/files#r1706246180

File: CIP-0XXX/README.md
Comment (rphair) [inline CIP-0XXX/README.md:434]:
```suggestion
### Information Diffusion Architecture
```
This is correctly sequenced with the _Rationale_ but according to CIP convention should also be a part of it  — rather than standing alone as an unconventional section — because it addresses design choices & alternatives.

**Review by rphair — COMMENTED** [review]
(intended to appear in sequence with above review, but GitHub ate it)

File: CIP-0XXX/README.md
Comment (rphair) [inline CIP-0XXX/README.md:521]:
```suggestion
## References
```
For consistency with other CIPs and CIP-0001 listed "optional sections": https://github.com/cardano-foundation/CIPs/blob/master/CIP-0001/README.md#structure

**Review by rphair — COMMENTED** [review]
(p.s.)

File: CIP-0XXX/README.md
Comment (ch1bo) [reply CIP-0XXX/README.md]: Done in https://github.com/cardano-scaling/CIPs/commit/a3677312b4fa951484ed769f3b69b6e90e93a742

File: CIP-0XXX/README.md
Comment (ch1bo) [reply CIP-0XXX/README.md]: Done in https://github.com/cardano-scaling/CIPs/commit/a3677312b4fa951484ed769f3b69b6e90e93a742

File: CIP-0XXX/README.md
Comment (ch1bo) [reply CIP-0XXX/README.md]: Done in https://github.com/cardano-scaling/CIPs/commit/a3677312b4fa951484ed769f3b69b6e90e93a742

File: CIP-0XXX/README.md
Comment (ch1bo) [reply CIP-0XXX/README.md]: Done in https://github.com/cardano-scaling/CIPs/commit/a3677312b4fa951484ed769f3b69b6e90e93a742

File: CIP-0XXX/README.md
Comment (ch1bo) [reply CIP-0XXX/README.md]: Done in https://github.com/cardano-scaling/CIPs/commit/a3677312b4fa951484ed769f3b69b6e90e93a742

File: CIP-0XXX/README.md
Comment (ch1bo) [reply CIP-0XXX/README.md]: Done in https://github.com/cardano-scaling/CIPs/commit/e154a146fb1ad42d98616814b28048b39d14ea71

File: CIP-0XXX/README.md
Comment (ch1bo) [reply CIP-0XXX/README.md]: Done in https://github.com/cardano-scaling/CIPs/commit/e154a146fb1ad42d98616814b28048b39d14ea71

File: CIP-0XXX/README.md
Comment (rphair) [inline CIP-0XXX/README.md:29]:
```suggestion
## Motivation: why is this CIP necessary?
```

File: CIP-0XXX/README.md
Comment (rphair) [inline CIP-0XXX/README.md:410]:
```suggestion
## Rationale: how does this CIP achieve its goals?
```

**Comment by rphair** [comment]
Tagging as `Unconfirmed` to reflect that this was positively received at CIP meeting # 94 but also that particulars like the `Category:` and CIP title are not definite yet.  This tag will also press it toward review at a future CIP meeting but the editors will wait for confirmation from the authors first that this is relatively complete & ready for community review.

File: CIP-0XXX/README.md
Comment (coot) [reply CIP-0XXX/README.md:43]: I think that should be made more clear in the accompanying text then.  Some people will expect that it's `cardano-node`.

**Comment by coot** [comment]
@rphair replying to https://github.com/cardano-foundation/CIPs/pull/876#issuecomment-2296775242

This opens many different applications, at least two of which were discussed internally recently: verification of tx-s making into the blockchain; distributing ledger peers in a decentralised way and many others.

File: CIP-0XXX/README.md
Comment (coot) [reply CIP-0XXX/README.md:43]: @ch1bo since we're proposing that `Message Producers` are communicating directly with themselves, that should be indicated in the diagram.  Only showing `n2n` communication of `cardano-node` suggests that will be used for message dessimination.

**Comment by coot** [comment]
@rphair I still think that my suggestion is better suited.  If I recall correctly this CIPs doesn't really mention what `Mithril` will be used for (nor the name `Decentralised Message Queue` suggests it).

**Comment by rphair** [comment]
@coot I'll make sure that every time this proposal is reviewed at CIP meetings that we talk about how the name & scope can be best related.  Also I hope to get more of the co-authors posting about this question.  It's above my level to assess so I would be inclined to follow whatever you & the authors would ultimately decide together.

**Comment by coot** [comment]
@rphair could you please invite me to the CIP meeting when this CIP is discussed?

File: CIP-0XXX/README.md
Comment (jpraynaud) [reply CIP-0XXX/README.md:43]: @coot the `Message Producers` will be, in the case of Mithril, the existing `Mithril signers` which will send signatures with the  `n2c` mini-protocols. 

In any case, I will update the document to explain what is meant by `Network Node`, `Message Producer` and `Message Consumer` in general and in particular for Mithril (referring to the `# Information Diffusion Architecture` section) :+1:

File: CIP-0XXX/README.md
Comment (jpraynaud) [reply CIP-0XXX/README.md:180]: Actually, this verification is rather done with the KES signature and Operation certificate embedded in the received message.

File: CIP-0XXX/README.md
Comment (jpraynaud) [reply CIP-0XXX/README.md]: @coot IIRC our conversations, you mentioned that the `Cardano node` would act as a proxy for the `Mithril network node` regarding the `n2c` communications. 

Is that correct or should we consider, in the case of Mithril, that the `Message Producer` will talk directly with the `Mithril network node` through a dedicated Unix socket?

**Comment by rphair** [comment]
@coot I think enough issues have come up in the discussion that this would be ready for Review now, so I've put it on the agenda for the next meeting (or can postpone by 2-week intervals at the same time of day, at your or other authors' request): https://hackmd.io/@cip-editors/96

I would first want to assign this a CIP number but normally we don't do that for a CIP without a `Category:`.  So the highest order for review (cc @Ryun1 @Crypto2099) would be for someone to summarise what's happening with the `Consensus` category for the CIP process so it can be assigned here, or possibly confirm `Network` as an alternative (https://github.com/cardano-foundation/CIPs/pull/876#discussion_r1705903729 - cc @ch1bo @jpraynaud @abailly-iohk + @bwbush from https://github.com/cardano-foundation/CIPs/pull/872#discussion_r1704141980).

We can also try to future-proof the title with respect to scope as per https://github.com/cardano-foundation/CIPs/pull/876#issuecomment-2309495449.

Meeting on CIP Discord (invite to channel: https://discord.gg/J8sGdCuKhs)

**Comment by jpraynaud** [comment]
> @rphair I still think that my suggestion is better suited. If I recall correctly this CIPs doesn't really mention what `Mithril` will be used for (nor the name `Decentralised Message Queue` suggests it).

@coot the original idea is to have a generic design for the `Decentralized Message Queue`. 

Mithril would be the first protocol using it, but as it is clearly mentioned in the `Motivation` section, other protocols (Leios, Peras, ...) will likely use it as well in the future.

File: CIP-0XXX/README.md
Comment (jpraynaud) [reply CIP-0XXX/README.md]: Done in https://github.com/cardano-scaling/CIPs/commit/6d64147f19c89557c217332ca1e79ceb5c76ba43

File: CIP-0XXX/README.md
Comment (jpraynaud) [reply CIP-0XXX/README.md]: Done in https://github.com/cardano-scaling/CIPs/commit/5e571686a45074a176cebf9823f9bc54f4e2583d

File: CIP-0XXX/README.md
Comment (jpraynaud) [reply CIP-0XXX/README.md]: Done in https://github.com/cardano-scaling/CIPs/commit/5b37ca2f2d71ec5c8fdf30620997b1812310e481 and https://github.com/cardano-scaling/CIPs/commit/74f4ddd72fef1d471b3f60ff402c01e79a2a945a

**Comment by rphair** [comment]
@jpraynaud I'm keeping an eye on Mermaid support on derived web sites that carry CIPs:
* https://github.com/cardano-foundation/cf-cip-frontend/issues/26
* https://github.com/cardano-foundation/developer-portal/issues/1312

File: CIP-0XXX/README.md
Comment (jpraynaud) [reply CIP-0XXX/README.md:149]: Done in https://github.com/cardano-scaling/CIPs/commit/23050341487317840577e3d9f7589fa881d1b026

File: CIP-0XXX/README.md
Comment (jpraynaud) [reply CIP-0XXX/README.md:78]: Done in https://github.com/cardano-scaling/CIPs/commit/73291a1af11c39ad8c22f4691196311081aa160b

File: CIP-0XXX/README.md
Comment (jpraynaud) [reply CIP-0XXX/README.md:43]: Done in https://github.com/cardano-scaling/CIPs/commit/1796de29ea0adc201a98a532d013723fec66a879

**Comment by rphair** [comment]
@jpraynaud @ch1bo I've marked all the individually posted conversations above as "resolved" wherever you've said they are fixed upstream, so we can get a fresh review after your last round of commits (especially if you can confirm you're done updating the branch for now).

In the future it will help if you yourselves can mark these conversations "resolved" when fixed in your working branch... since making those changes won't automatically resolve the conversations in this PR.  That way we can keep this thread cleaner for identifying & discussing new review points. 🙏

File: CIP-0XXX/README.md
Comment (Ryun1) [reply CIP-0XXX/README.md:4]: Whilst we are [Rethinking Project Enlisting](https://github.com/cardano-foundation/CIPs/issues/898)
Im happy to move and assign `Networking` for now, with the understanding that we are working on how we navigate project enlistment in a post-chang world.

Comment (rphair) [reply CIP-0XXX/README.md:4]: @coot @jpraynaud @abailly would you prefer `Network` or `Networking`?

Comment (jpraynaud) [reply CIP-0XXX/README.md:4]: `Network` seems fine to me :+1:

Comment (rphair) [reply CIP-0XXX/README.md:4]: thanks @coot @jpraynaud @abailly - I've amended it above (from `Consensus`) and we'll collect similar proposals under the `Network` category term (unless `Consensus` is better applicable for those)... though like we said at yesterday's meeting there aren't likely to be many.  cc @Ryun1

Comment (jpraynaud) [reply CIP-0XXX/README.md]: Done in https://github.com/cardano-scaling/CIPs/commit/e44e9ac1315264db6f8d9f9f2bedb9d40bcae38c :heavy_check_mark:

**Comment by rphair** [comment]
@jpraynaud @abailly-iohk - applying freshly created label https://github.com/cardano-foundation/CIPs/labels/Category%3A%20Network - please chime in with this issue with any feelings about how you see the CIP process participation for this category developing:
- https://github.com/cardano-foundation/CIPs/issues/898

In your opinion, would this be ready for Review at the next CIP meeting? (https://hackmd.io/@cip-editors/98)  If not, please post something about how this proposal has been progressing & what we're waiting for.

**Comment by rphair** [comment]
p.s. to https://github.com/cardano-foundation/CIPs/pull/876#issuecomment-2396476898 @jpraynaud @abailly-iohk - another advantage to putting this up for Review again is that we could assign it a CIP number, especially now that questions about scope & categorisation seem to have been answered.  RSVP

File: CIP-0XXX/README.md
Comment (jpraynaud) [reply CIP-0XXX/README.md]: @coot I made clear that the node being called by the n2c mini-protocols is the network node and not the core node in https://github.com/cardano-scaling/CIPs/commit/0f4262e50eeb610054ccf9378644d44f10aa600e.

File: CIP-0XXX/README.md
Comment (jpraynaud) [reply CIP-0XXX/README.md]: Fixed in https://github.com/cardano-scaling/CIPs/commit/0f4262e50eeb610054ccf9378644d44f10aa600e.

File: CIP-0XXX/README.md
Comment (jpraynaud) [reply CIP-0XXX/README.md]: Following our conversations with @coot and @ch1bo, we will name this external process node `Decentralized Message Queue` or `DMQ` node. I applied the modification in https://github.com/cardano-scaling/CIPs/commit/e3ac2cdcdad97362839bd3c9e066b2c4b54f8fea.

**Comment by jpraynaud** [comment]
> thanks @jpraynaud - I've put it on the agenda again & hope you can attend: https://hackmd.io/@cip-editors/98

Thanks @rphair, I'll be attending the next CIP meeting :+1:.

File: CIP-0XXX/README.md
Comment (rphair) [inline CIP-0XXX/README.md:2]:
```suggestion
CIP: 137
```

**Review by rphair — COMMENTED** [review]
Thanks @jpraynaud & team for such a close cooperation with the editors & community on this one: well confirmed as candidate in today's CIP meeting & therefore marking `Last Check` & therefore ready to merge unless any problems are identified in the next 2 weeks or so.

Please also update to the new CIP number by renaming the containing directory & change the proposal link in this PR's first posting accordingly. 🎉

Personally I believe this has passed all documentation requirements & peer review and will  ✅ it as soon as this number assignment is made in the document:

File: CIP-0XXX/README.md
Comment (jpraynaud) [reply CIP-0XXX/README.md]: Done in https://github.com/cardano-scaling/CIPs/commit/fe7eb48f9ddd12e226366f84871cfd019fa96559

File: CIP-0137/README.md
Comment (perturbing) [inline CIP-0137/README.md:31]: I believe the motivation section could benefit from a more generalized framing before diving into the specific application of Mithril. As it currently stands, the document immediately presents Mithril as the driving example, which might confuse readers who are looking for a broader understanding of the need for a decentralized message queue.

I would suggest starting with a more generalized motivation that emphasizes the overarching need for decentralized message diffusion across various protocols in the Cardano ecosystem. The benefits of using the existing Cardano infrastructure for this purpose—such as efficiency, security, and reduced development overhead—should be highlighted first. Mithril can then be introduced as a specific use case, perhaps noting that it will be used as an illustrative example throughout the document.

While I understand Mithril will be a fundamental first user of this DMQ, it's important that the document doesn't feel too Mithril-centric.

**Comment by Quantumplation** [comment]
I might have missed it in my quick read, but does this deal with how long messages should be retained for diffusion by each node?  even if it's not enforceable at the protocol level, if this is meant to be a generic mechanism, it'd be useful to know the bounds on availability for people intending to use this mechanism.

For example, does the node only propagate the messages to the *currently* connected peers? does it keep them available for download for 24 hours? does it keep them *permanently*? etc.

**Comment by jpraynaud** [comment]
> de only propagate the messages to the _currently_ connected peers? does it keep them available for download for 24 hours? does it keep them _permanently_? etc.

Thanks @Quantumplation for raising this question.

The messages will have a lifetime configured a the topic level: messages will be considered invalid after that lifetime has elapsed, and will neither be diffused to other network nodes via the submission n2n submission mini-protocol nor accessible by the n2c notification mini-protocol.

I have detailed this invalidation mechanism and some figures for the overall extra storage requirements in the case of Mithril in the commit https://github.com/cardano-scaling/CIPs/commit/f97083ee32a1a16cf230d6198f6f800232475954.

File: CIP-0137/README.md
Comment (jpraynaud) [reply CIP-0137/README.md:31]: Thanks @perturbing for this feedback. I will adjust the motivation section in that manner shortly :+1:

Comment (jpraynaud) [reply CIP-0137/README.md]: @perturbing I have updated the Motivation section to be less Mithril-centric in the commit https://github.com/cardano-scaling/CIPs/commit/7b304a559c4d59a7722782fcfb585f946fcdd5bc

Comment (rphair) [reply CIP-0137/README.md]: thanks @jpraynaud - I think https://github.com/cardano-foundation/CIPs/pull/876/commits/7b304a559c4d59a7722782fcfb585f946fcdd5bc does a great job of putting this CIP into context without being too specific with respect to Milthril.

**Review by Crypto2099 — APPROVED** [review]
All community feedback has been sufficiently addressed, ready to merge in.

**Pull request merged**
